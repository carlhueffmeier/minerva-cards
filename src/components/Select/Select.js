import React, { Component } from 'react';
import PropTypes from 'prop-types';

const optionType = PropTypes.shape({
  id: PropTypes.string.isRequired // every option needs an id
});

export default class Select extends Component {
  static propTypes = {
    title: PropTypes.string.isRequired, // e.g. "select deck"
    options: PropTypes.arrayOf(optionType).isRequired, // possible values
    value: PropTypes.string.isRequired, // current value
    display: PropTypes.func.isRequired, // specifies how to display an option
    onChange: PropTypes.func.isRequired, // receives the id of new option
    onAdd: PropTypes.func // optional handler to add new option
  };

  handleChange(event) {
    this.props.onChange(event.target.value);
  }

  render() {
    const { title, options, value, display, onAdd } = this.props;

    return (
      <div>
        <label>
          <span className="select__title">{title}</span>
          <select
            name="type"
            value={value}
            onChange={this.handleChange.bind(this)}
          >
            {options.map(option => (
              <option
                className="select__option"
                key={option.id}
                value={option.id}
              >
                {display(option)}
              </option>
            ))}
          </select>
        </label>
        {onAdd && <button onAdd={onAdd}>+</button>}
      </div>
    );
  }
}
