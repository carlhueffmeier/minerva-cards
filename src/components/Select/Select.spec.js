import React from 'react';
import { shallow } from 'enzyme';
import Select from './Select';

describe('Select', () => {
  const props = {
    title: `Choose Deck`,
    options: [{ id: `1`, name: `Default` }, { id: `2`, name: `Kanji` }],
    value: `2`,
    display: option => option.name,
    onChange: newValue => console.log(newValue)
  };

  const select = shallow(<Select {...props} />);

  it('renders properly', () => {
    expect(select).toMatchSnapshot();
  });

  it('displays the title', () => {
    expect(select.find(`.select__title`).text()).toEqual(props.title);
  });

  it('renders the right number of options', () => {
    expect(select.find(`.select__option`)).toHaveLength(2);
  });
});
