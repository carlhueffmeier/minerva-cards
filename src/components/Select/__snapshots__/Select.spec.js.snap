// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Select
    display={[Function]}
    onChange={[Function]}
    options={
        Array [
            Object {
              "id": "1",
              "name": "Default",
            },
            Object {
              "id": "2",
              "name": "Kanji",
            },
          ]
    }
    title="Choose Deck"
    value="2"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label>
          <span
                    className="select__title"
          >
                    Choose Deck
          </span>
          <select
                    name="type"
                    onChange={[Function]}
                    value="2"
          >
                    <option
                              className="select__option"
                              value="1"
                    >
                              Default
                    </option>
                    <option
                              className="select__option"
                              value="2"
                    >
                              Kanji
                    </option>
          </select>
</label>,
        undefined,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="select__title"
>
              Choose Deck
</span>,
            <select
              name="type"
              onChange={[Function]}
              value="2"
>
              <option
                            className="select__option"
                            value="1"
              >
                            Default
              </option>
              <option
                            className="select__option"
                            value="2"
              >
                            Kanji
              </option>
</select>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Choose Deck",
              "className": "select__title",
            },
            "ref": null,
            "rendered": "Choose Deck",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  className="select__option"
                  value="1"
>
                  Default
</option>,
                <option
                  className="select__option"
                  value="2"
>
                  Kanji
</option>,
              ],
              "name": "type",
              "onChange": [Function],
              "value": "2",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "host",
                "props": Object {
                  "children": "Default",
                  "className": "select__option",
                  "value": "1",
                },
                "ref": null,
                "rendered": "Default",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": "2",
                "nodeType": "host",
                "props": Object {
                  "children": "Kanji",
                  "className": "select__option",
                  "value": "2",
                },
                "ref": null,
                "rendered": "Kanji",
                "type": "option",
              },
            ],
            "type": "select",
          },
        ],
        "type": "label",
      },
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label>
            <span
                        className="select__title"
            >
                        Choose Deck
            </span>
            <select
                        name="type"
                        onChange={[Function]}
                        value="2"
            >
                        <option
                                    className="select__option"
                                    value="1"
                        >
                                    Default
                        </option>
                        <option
                                    className="select__option"
                                    value="2"
                        >
                                    Kanji
                        </option>
            </select>
</label>,
          undefined,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="select__title"
>
                Choose Deck
</span>,
              <select
                name="type"
                onChange={[Function]}
                value="2"
>
                <option
                                className="select__option"
                                value="1"
                >
                                Default
                </option>
                <option
                                className="select__option"
                                value="2"
                >
                                Kanji
                </option>
</select>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Choose Deck",
                "className": "select__title",
              },
              "ref": null,
              "rendered": "Choose Deck",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option
                    className="select__option"
                    value="1"
>
                    Default
</option>,
                  <option
                    className="select__option"
                    value="2"
>
                    Kanji
</option>,
                ],
                "name": "type",
                "onChange": [Function],
                "value": "2",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Default",
                    "className": "select__option",
                    "value": "1",
                  },
                  "ref": null,
                  "rendered": "Default",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": "2",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Kanji",
                    "className": "select__option",
                    "value": "2",
                  },
                  "ref": null,
                  "rendered": "Kanji",
                  "type": "option",
                },
              ],
              "type": "select",
            },
          ],
          "type": "label",
        },
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
